File: tools/test_canonical_hash.py
#!/usr/bin/env python3
"""Python side of the canonical-hash byte-equivalence harness.

For every fixture in tests/cross_canonical/, compute
canonical_hash.canonical_sha256 and assert it equals the value pinned
in tests/cross_canonical/expected.json. The OCaml side
(sdk/test/test_canonical_hash.ml) does the same against the same
sidecar. Both passing on a given fixture means OCaml's
Yojson.Safe.to_string + Codec.normalize and Python's json.dumps +
_canonicalize produce byte-identical output on that input — the
locked invariant every cross-document hash check downstream depends
on.

Pure stdlib; runs in CI's schemas job. Exits non-zero on the first
failing fixture so the failure is visible without scrolling.
"""
from __future__ import annotations

import json
import sys
from pathlib import Path

# Allow running directly (python tools/test_canonical_hash.py) or as a
# module — both must find canonical_hash next to this file.
sys.path.insert(0, str(Path(__file__).resolve().parent))
import canonical_hash as ch  # noqa: E402


ROOT = Path(__file__).resolve().parent.parent
CROSS = ROOT / "tests" / "cross_canonical"


def main() -> int:
    expected = json.loads((CROSS / "expected.json").read_text())
    fixtures = sorted(
        p.name for p in CROSS.glob("*.json") if p.name != "expected.json"
    )
    failures = 0
    for name in fixtures:
        want = expected.get(name)
        if want is None:
            print(f"FAIL  {name}  (no entry in expected.json)")
            failures += 1
            continue
        obj = json.loads((CROSS / name).read_text())
        got = ch.canonical_sha256(obj)
        if got == want:
            print(f"PASS  {name}  {got}")
        else:
            print(f"FAIL  {name}")
            print(f"  expected: {want}")
            print(f"  got:      {got}")
            failures += 1
    if failures:
        print(f"\n{failures}/{len(fixtures)} fixtures failed — "
              f"OCaml and Python canonical-hash byte streams have drifted. "
              f"DO NOT update expected.json to make this pass; investigate "
              f"the divergence (see Codec.canonical_bytes docstring).",
              file=sys.stderr)
        return 1
    print(f"\nAll {len(fixtures)} fixtures pass — OCaml<->Python "
          f"canonical-hash byte agreement preserved.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdk/test/test_canonical_hash.ml
(** OCaml side of the canonical-hash byte-equivalence harness.

    For every fixture in [tests/cross_canonical/], compute
    [Hash.canonical_sha256] and assert it equals the value pinned in
    [tests/cross_canonical/expected.json]. The Python side
    ([tools/test_canonical_hash.py]) does the same against the same
    sidecar. Both agreeing = OCaml's [Yojson.Safe.to_string +
    Codec.normalize] and Python's [json.dumps + _canonicalize] are
    producing byte-identical output, which is the locked invariant
    every cross-document hash check downstream depends on.

    Test failure here means OCaml drifted from the sidecar.
    [tools/test_canonical_hash.py] failing means Python did. If both
    drift but agree with each other against the sidecar, the harness
    surfaces it as TWO failing tests against expected.json — never
    a silent same-pass on a divergent format. *)

open Proof_broker

(* The dune build tree puts test_canonical_hash.exe at
   [_build/default/sdk/test/test_canonical_hash.exe], from which the
   repo root is four levels up. Mirrors test_round_trip.ml's path. *)
let cross_dir =
  Filename.concat (Sys.getcwd ()) "../../../../tests/cross_canonical"

let read_file path =
  let ic = open_in path in
  let n = in_channel_length ic in
  let s = really_input_string ic n in
  close_in ic;
  s

let expected_table () : (string * string) list =
  let path = Filename.concat cross_dir "expected.json" in
  match Yojson.Safe.from_file path with
  | `Assoc pairs ->
    List.filter_map (fun (k, v) ->
      if String.length k > 0 && k.[0] = '_' then None
      else match v with `String s -> Some (k, s) | _ -> None
    ) pairs
  | _ -> failwith "expected.json is not a top-level object"

let test_fixture (fixture, expected_sha) () =
  let path = Filename.concat cross_dir fixture in
  let j = Yojson.Safe.from_file path in
  let actual = Hash.canonical_sha256 j in
  Alcotest.(check string)
    (Printf.sprintf "canonical_sha256 for %s" fixture)
    expected_sha actual

let () =
  let cases =
    List.map (fun (fixture, sha) ->
      Alcotest.test_case fixture `Quick (test_fixture (fixture, sha)))
      (expected_table ())
  in
  Alcotest.run "canonical_hash" [
    "byte-equivalence vs tests/cross_canonical/expected.json", cases;
  ]
